experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +453,184 @@ impl LocaleDisplayNamesFormatter {
+    locale: &Locale,
+    longest_matching_subtag: LongestMatchingSubtag,
+    locale_dn_formatter: &'a LocaleDisplayNamesFormatter,
+) -> &'a str {
+    let LocaleDisplayNamesFormatter {
+        options,
+        locale_data,
+        language_data,
+        ..
+    } = locale_dn_formatter;
+
+    let lang_id: LanguageIdentifier = match longest_matching_subtag {
+        LongestMatchingSubtag::LangRegion => (locale.id.language, None, locale.id.region).into(),
+        LongestMatchingSubtag::LangScript => (locale.id.language, locale.id.script, None).into(),
+        LongestMatchingSubtag::Lang => locale.id.language.into(),
+    };
+
+    // Check if the key exists in the locale_data first.
+    // Example: "en_GB", "nl_BE".
+    let mut ldn = match options.style {
+        Some(Style::Short) => locale_data
+            .get()
+            .short_names
+            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
+        Some(Style::Long) => locale_data
+            .get()
+            .long_names
+            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
+        Some(Style::Menu) => locale_data
+            .get()
+            .menu_names
+            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse()),
+        _ => None,
+    }
+    .or_else(|| {
+        locale_data
+            .get()
+            .names
+            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse())
+    });
+
+    // At this point the key should exist in the language_data.
+    // Example: "en", "nl", "zh".
+    if ldn.is_none() {
+        ldn = match options.style {
+            Some(Style::Short) => language_data
                 .get()
                 .short_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
-            Some(Style::Long) => self
-                .language_data
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
+            Some(Style::Long) => language_data
                 .get()
                 .long_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
-            Some(Style::Menu) => self
-                .language_data
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
+            Some(Style::Menu) => language_data
                 .get()
                 .menu_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
             _ => None,
         }
         .or_else(|| {
-            self.language_data
+            language_data
                 .get()
                 .names
-                .get(&locale.id.language.into_tinystr().to_unvalidated())
+                .get(&lang_id.language.into_tinystr().to_unvalidated())
         });
+    }
+    // Throw an error if the LDN is none as it is not possible to have a locale string without the language.
+    return ldn.expect("cannot parse locale displayname.");
```

Comment:
Please make this do the fallback behavior instead of panic

Reply:
Done